Nickvision.Aura/Network/NetworkMonitor.cs

```diff
@@ -0,0 +1,141 @@
+using System;
+using System.Net.NetworkInformation;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using Tmds.DBus;
+
+namespace Nickvision.Aura.Network;
+
+/// <summary>
+/// Network state monitor
+/// </summary>
+public class NetworkMonitor : IDisposable
+{
+    public event Action<NetworkState>? _stateChanged;
+    
+    /// <summary>
+    /// Occurs when network state is changed
+    /// </summary>
+    public event Action<NetworkState> StateChanged
+    {
+        add
+        {
+            _stateChanged += value;
+            Task.Run(async () => _stateChanged.Invoke(await GetStateAsync()));
+        }
+
+        remove
+        {
+            _stateChanged -= value;
+        }
+    }
+    
+    private bool _disposed;
+    private Connection? _dbusConnection;
+    private INetworkManager? _networkManager;
+    private string[]? _networkAddresses;
+    
+    /// <summary>
+    /// Construct NetworkMonitor
+    /// </summary>
+    public NetworkMonitor()
+    {
+        _disposed = false;
+        if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+        {
+            #pragma warning disable CS4014
+            SetupNetworkManager();
```

Comment:
I'd move the code from the method directly here since it's just used here

Reply:
we can't use await in Constructor, but we need it, otherwise we can't catch async exceptions from dbus

Reply:
ah didn't see it was async....
um this is a problem then. we should do this then:

```csharp
private NetworkMonitor()
{
    _disposed = false;
}

public static async Task<NetworkMonitor> NewAsync()
{
    var netmon = new NetworkMontior();
    if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
    {
        try
        {
            netmon._dbusConnection = new Connection(Address.System);
            await netmon._dbusConnection.ConnectAsync();
            netmon._networkManager = _dbusConnection.CreateProxy<INetworkManager>("org.freedesktop.NetworkManager", new ObjectPath("/org/freedesktop/NetworkManager"));
            await netmon._networkManager.WatchStateChangedAsync(change => _stateChanged?.Invoke(change));
        }
        catch
        {
            netmon._networkManager = null;
            netmon.SetupPing();
        }
    }
    else
    {
        netmon.SetupPing();
    }
    return netmon;
}
```

Reply:
what we can do is to create a static sync method that will call an empty constructor and then setup, something line `NetworkMonitor.Init()`

Reply:
ah your comment just appeared